src/storage/store/engine.rs

```diff
@@ -1303,7 +1260,88 @@ impl ShardEngine {
                 // This should never happen
                 panic!("No body in event");
             }
+        };
+        Ok((inserts, deletes))
+    }
+
+    fn update_trie(
+        &mut self,
+        ctx: &merkle_trie::Context,
+        event: &proto::HubEvent,
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<(), EngineError> {
+        let now = std::time::Instant::now();
+        let (inserts, deletes) = self.get_trie_update_keys(event)?;
+
+        for key in inserts {
+            self.stores
+                .trie
+                .insert(ctx, &self.db, txn_batch, vec![&key])?;
+        }
+        for key in deletes {
+            self.stores
+                .trie
+                .delete(ctx, &self.db, txn_batch, vec![&key])?;
+        }
+
+        let elapsed = now.elapsed();
+        self.time_with_shard("update_trie_time_us", elapsed.as_micros() as u64);
+        Ok(())
+    }
+
+    /// Insert a vec of events all at once into the trie
+    #[allow(dead_code)]
+    fn update_trie_batch(
+        &mut self,
+        ctx: &merkle_trie::Context,
+        events: &Vec<proto::HubEvent>,
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<(), EngineError> {
+        let now = std::time::Instant::now();
+
+        let mut all_inserts = HashSet::new();
+        let mut all_deletes = HashSet::new();
+
+        for event in events {
+            let (inserts, deletes) = self.get_trie_update_keys(event)?;
+
+            // Merge the inserts and deletes
```

Comment:
This looks fine, but might we worth adding a test with a few scenarios unless it's covered by replication tests?

Reply:
The replication tests will cover it (All the replication code will go through this, and there are tests in the other PR that cover this function)